### ISSUE TITLE
Convert NO-B43F OP1 and OP2 readings to concentration data

### ISSUE THREAD

**Issue by ruihangdu** [waagsociety/making-sensor#1]
Hello,
  How did you convert the voltage difference to NO<sub>2</sub> concentration? Much appreciated!

Ray

**Comment by sbocconi** [comment]
Hi Ray,

There are 2 methods you can use, or at least that were known to us. The first is to use the parameters of the factory to calculate the concentration, but there are some assumptions which I am not aware now and we have not followed that path.

The second is to assume that NO2 = a OUT1 + b OUT2 + c Temp + d Humidity + offset

Meaning that the NO2 concentration is a linear combination of the 2 outputs of the Alphasense,  the temperature and the humidity (we had another sensor measuring those), and an offset.

To calculate a,b,c,d, and offset (calibration)  you need to have reliable data and apply linear regression. In our case we put all the sensors near a measuring station of the Dutch Health Organisation (GGD) to get reliable data about the NO2 concentration, and then did the math.

Basically if you cannot access reliable data to perform the calibration, you need to use the parameters from the factory, but I would not suggest this,

I hope it makes sense to you.

Regards,

	Stefano

**Comment by P992** [comment]
Can you explain better how you calculate a, b, c, d and offset? Do You have an excel that make it? Thank you so much for your share

**Comment by sbocconi** [comment]
We used some Python code to perform the linear regression. I am not the owner of the code, so I cannot share it without asking, if this is an option for you I can ask the author if he allows to share it and possibly upload it here.

**Comment by P992** [comment]
Yes thank you. If is possible share it. Thank you again

Il lun 27 mag 2019, 10:27 Stefano Bocconi <notifications@github.com> ha
scritto:

**Comment by P992** [comment]
Hy if is not possibile to share the original code can you share a tutorial that explain how to calculate No2 without use alphasense aglorithm ? thank you

**Comment by Tessatino** [comment]
Hi, so we used a python code to do the following:
- Create a pandas Dataframe with the data (this includes 'true' no2 data from official measurements, and from our own sensor: no2a, no2b, temp, humidity)
- Then the Ordinary Least Squares '.ols' statsmodel function is used to calculate the parameters of the formula, by getting the formula and the data as input. See more information about the .ols function here: [http://www.statsmodels.org/dev/examples/notebooks/generated/ols.html](http://www.statsmodels.org/dev/examples/notebooks/generated/ols.html)
- The parameters can then be obtained by the function '.params', for example: a = results.params['no2a']

Hope this helps you further! If it is possible for us to share the code we will let you know.

**Comment by P992** [comment]
Thank you for help
